go/vt/orchestrator/logic/topology_recovery.go

```diff
@@ -1998,13 +2018,6 @@ func GracefulMasterTakeover(clusterName string, designatedKey *inst.InstanceKey,
 	if topologyRecovery == nil {
 		return nil, nil, fmt.Errorf("GracefulMasterTakeover: recovery attempted but with no results. This should not happen")
 	}
-	if topologyRecovery.SuccessorKey == nil {
-		// Promotion failed. Undo.
-		log.Infof("GracefulMasterTakeover: Invoking tabletUndoDemoteMaster on %+v", clusterMaster.Key)
-		if err := TabletUndoDemoteMaster(clusterMaster.Key); err != nil {
-			log.Errore(err)
-		}
-	}
```

Comment:
I'm probably unable to follow the logic; can't there be a situation where `topologyRecovery.SuccessorKey == nil`?

Reply:
I moved this inside `recoverDeadmaster`. The reason is because this function does not `LockShard`, and was causing race conditions against other actors.

Adding a `LockShard` is problematic because it eventually invokes `recoverDeadMaster`, which can be called through other code paths.

It's still not perfect. Barely functional. I'm thinking of adding a couple of new flows that can handle `LockShard` more cleanly.

Reply:
:+1: thank you